common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+     * to return immediately, but the mapper function cannot be applied before this future has completed.
+     * <p>
+     * Effectively, this method behaves similar to this serial code, except asynchronously and with proper exception
+     * and cancellation handling:
+     * <pre>{@code
+     * V x = future.sync().getNow();
+     * Future<R> y = mapper.apply(x);
+     * R result = y.sync().getNow();
+     * }</pre>
+     * <p>
+     * If the given future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but the given future will be unaffected.
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R>    The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    public static <V, R> Future<R> flatMap(Future<V> future, Function<V, Future<R>> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+        Promise<R> promise = future.executor().newPromise();
+        future.addListener(new FlatMapper<>(promise, mapper));
+        if (!future.isSuccess()) {
+            // Propagate cancellation if future is either incomplete or failed.
+            // Failed means it could be cancelled, so that needs to be propagated.
+            promise.addListener(future, PromiseNotifier::propagateCancel);
+        }
+        return promise;
+    }
+
+    private Futures() {
+    }
+
+    @SuppressWarnings("unchecked")
+    static <R> FutureContextListener<Promise<R>, Object> passThrough() {
+        return (FutureContextListener<Promise<R>, Object>) (FutureContextListener<?, ?>) PASS_THROUGH_CONSTANT;
+    }
+
+    static <A, B> void propagateUncommonCompletion(Future<? extends A> completed, Promise<B> recipient) {
+        if (completed.isCancelled()) {
+            // Don't check or log if cancellation propagation fails.
+            // Propagation goes both ways, which means at least one future will already be cancelled here.
+            recipient.cancel(false);
+        } else {
+            Throwable cause = completed.cause();
+            tryFailure(recipient, cause, LOGGER);
+        }
+    }
+
+    private static class PassThrough<R> implements FutureContextListener<Promise<R>, Object> {
+        @Override
+        public void operationComplete(Promise<R> recipient, Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    recipient.trySuccess(result);
+                } catch (RuntimeException e) {
```

Comment:
I think for completing async control flow, catching `Throwable` is a better approach than just `RuntimeException`. 
In this case, if we miss any error here, we would never complete the `promise` which may be very hard to debug.